Block Supports: Add border color, style and width support

aaronrobertshaw opened the pull request:
Fixes: #29616

## Description
This PR extends the border related block support to include color, style and width options.

These border options are disabled by default.

## How has this been tested?

Manually with a couple of small updates to existing unit tests.

#### Testing Instructions:

1. Enable custom border color, radius, style and width in your theme.json file.
2. Create a new post and add a group block to it
3. Select the group block and confirm the "Border settings" panel displays in the sidebar
4. Expand the "Border settings" panel and try editing each of the border properties. 
5. Confirm the block updates as expected in the editor, save and then confirm correct classes and styles on the frontend.
6. Open the Site Editor then its global styles sidebar. Switch to the "By block type" tab and expand the Group block panel.
7. Confirm that the border support features you have enabled are present. 
8. Test changing the global styles for the group block's borders.

#### Unit Tests:
- `npm run test-unit:watch packages/block-editor/src/hooks/test/style.js`
- `wp-env run phpunit "phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php"`

## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/60436221/112101007-b7164000-8bf1-11eb-9bb0-168af504cf15.gif" width="400" />

## Types of changes
New feature

## Next steps

- [ ] Make any required UI refinements
- [ ] Confirm approach to border color support
- [ ] Surface the border color option under the colors panel as well if desired (https://github.com/WordPress/gutenberg/issues/27331#issuecomment-793799694)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Add border color, style and width support

aaronrobertshaw commented:
@nosolosw I've made a start on adding the extra properties to the border block support. It would be great if you could spare a few moments to sanity check the approach taken for the border colors. Thanks in advance!

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+			},
+		};
+
+		const newNamedColor = colorObject?.slug ? colorObject.slug : undefined;
```
`colorObject?.slug` will return `undefined' if there is no 'slug' property. I believe this ternary is not needed here.

aaronrobertshaw replied:
I was under the impression the colors were coming directly from the theme.json. Given it could be supplied by a theme developer and the slug accidentally left as an empty string it seemed prudent to ensure undefined. It's also consistent with `color.js` and `with-colors.js` at present.

Happy to change it of course if it should be simplified to `colorObject?.slug`

grzim replied:
ah, if this is a case, I think it would be good to add a comment as this piece of code could be misleading. Alternatively we can write:
```js
const namedBorderColor = colorObject?.slug ? {borderColor:  colorObject.slug} : {} 

setAttributes( {
			style: cleanEmptyObject( newStyle ),
			...namedBorderColor
		} );
```
thanks to which we will not have a property with `undefined` assigned.

aaronrobertshaw replied:
I've added a comment regarding enforcing undefined in case a theme gives us an empty slug value in the color preset. Hopefully that might assist in making things clearer.

The suggested code snippet would actually introduce a bug I'm afraid. We need to set the attribute to undefined to clear any previous named color selection.

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+import { cleanEmptyObject } from './utils';
+
+const BORDER_COLOR_SUPPORT_KEY = 'color';
+const EMPTY_ARRAY = [];
```
introducing such a variable is not a good idea. When it will be assigned in multiple places it will always point to the same reference, which could cause strange problems. It is better to create empty arrays inline.

gziolo replied:
It depends. If it is passed later as a prop, then it prevents unnecessary re-renders. I see the same pattern used with different color hooks. Unfortunately it isn't documented.

aaronrobertshaw replied:
I've updated this to pass the colors and gradients to the `ColorGradientControl` via props. This means the control no longer needs to retrieve the colors a second time. The current approach with `EMPTY_ARRAY` does now also prevent those unnecessary re-renders as mentioned.

grzim replied:
I see. Normally in such cases, when we want to avoid rerenders we would use useRef. 
This will also work as expected, but I think it would be good to add a comment to avoid confusion.

aaronrobertshaw replied:
Thanks, I've added a comment.

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
+			},
```
we can use loadash here to simplify the logic a little
```js

_.merge(style, {border: {color: colorObject?.slug && value }})
```

loadash will handle case when colorObject?.slug returns undefined

aaronrobertshaw replied:
I'm reluctant to increase use of lodash when this is [consistent with the existing color hook](https://github.com/WordPress/gutenberg/blob/update/border-block-support/packages/block-editor/src/hooks/color.js#L248). 

Happy to change if others agree.

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+ */
+export function hasBorderColorSupport( blockType ) {
+	const support = getBlockSupport( blockType, BORDER_SUPPORT_KEY );
+	return !! ( true === support || support?.color );
```
This code implies that `getBlockSupport` can return either object. Is this a case?

aaronrobertshaw replied:
Yes. All border support can be opted into by simply setting the `supports.__experimentalBorder` property to true. Alternatively, individual features ( color, radius, style, and width ) can be opted into.

aaronrobertshaw replied:
I've refactored all the individual feature checks to use a single more generic function exported from `border.js`

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+
+	// Allow blocks to specify default value if needed.
+	if ( ! settings.attributes.borderColor ) {
+		Object.assign( settings.attributes, {
```
changing the variable passed as an argument is a bad practice. It is better to return a new object with updated value

aaronrobertshaw replied:
Thanks, I've updated this to return a new settings object including the new `borderColor` attribute definition.

grzim commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,197 @@
+		[ borderColorClass ]: !! borderColorClass,
+	} );
+
+	props.className = newClassName ? newClassName : undefined;
```
what is the intention behind this check?

aaronrobertshaw replied:
It is to prevent an empty string being assigned to `className` when the named border colors are being toggled on and off while no other classes are present on the block.

grzim replied:
Wouldn't it be more straightforward in this way:
```js
if(newClassName)
   props.className = newClassName;
```
?

aaronrobertshaw replied:
Unfortunately not. It would fail to clear the previous border color classes from the `className` attribute when they were the only ones present.

Take a situation like:
1. Block only has border color classes in its `className`
2. User clears border color selection
3. The new calculated class name in `newClassName` is correctly empty.
4. In the above snippet, it would be falsey and so not clear the `className` prop as required.

I've added another comment to the code to explain this a little more. Hopefully that helps some.

grzim commented on packages/block-editor/src/hooks/border-style.js:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	const onChange = ( newBorderStyle ) => {
+		const newStyleAttributes = {
```
here `merge` can be used as well

aaronrobertshaw replied:
It could be. I'll make this change if it is [decided we want to increase lodash usage](https://github.com/WordPress/gutenberg/pull/30124#discussion_r600094420).

grzim commented on packages/block-editor/src/hooks/border-style.js:
```diff
@@ -0,0 +1,70 @@
+ * @param  {string|Object} blockType Block name or block type object.
+ * @return {boolean}                 Whether there is support.
+ */
+export const hasBorderStyleSupport = ( blockType ) => {
```
this function is doubled

aaronrobertshaw replied:
I've refactored all the individual feature checks to use a single more generic function exported from `border.js`

grzim commented on packages/block-editor/src/hooks/border-width.js:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	const onChange = ( newWidth ) => {
+		const newStyle = {
```
we can use `merge` here

grzim commented on packages/block-editor/src/hooks/border-width.js:
```diff
@@ -0,0 +1,79 @@
+ * @param  {string|Object} blockType Block name or block type object.
+ * @return {boolean}                 Whether there is support.
+ */
+export const hasBorderWidthSupport = ( blockType ) => {
```
this is the 3rd appearance of the same function. We can export it to a separate file, so it can be used among all 3 different files.

aaronrobertshaw replied:
I've refactored all the individual feature checks to use a single more generic function exported from `border.js`

aaronrobertshaw committed:
Pass colors and gradients to border color control

Passing these props prevents the need for the ColorGradientControl to retrieve the colors a second time.

aaronrobertshaw committed:
Return new object instead of mutating passed settings

aaronrobertshaw committed:
Refactor individual feature support checks

aaronrobertshaw commented:
Thanks for the time and effort on the review @grzim. Appreciate it 👍

aaronrobertshaw committed:
Add support for skipping serialization of border color

aaronrobertshaw committed:
Add comment about enforcing undefined for color slug

aaronrobertshaw committed:
Add comment explaining EMPTY_ARRAY avoiding rerenders

aaronrobertshaw committed:
Add clarifying comment around clearing `className`

aaronrobertshaw commented:
I've rebased this PR and added skipping serialization of border support properties. There have also been a few inline comments added to help clarify the intent of the code.

stacimc reviewed (commented):
This tests well for me in both the editor and on the frontend, with various permutations of the available settings. I also tested enabling only a subset of the color/style/width/radius options in the `theme.json`, omitting the color support for example.

LGTM, this would be a great addition 🎉

apeatling reviewed (approved):
Looking really good, thanks for your work on this one. Tested as per instructions and it worked well. ♥️

aaronrobertshaw committed:
Add border block support to global styles sidebar

aaronrobertshaw commented:
A new border panel has been added to the Global Styles sidebar and the PRs testing instructions updated. This side of the changes could do with closer inspection and testing.

ramonjd commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -0,0 +1,202 @@
+			...style,
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
```
I'm just testing this PR and, for me, the border color panel doesn't add the border color inline style to the root element. The `has-{color}-border-color` class is present however.

The result is that the border color doesn't change when I select a new border color. 

Strangely, changing the text color modifies the border color, though I can't see why in the computed styles.

![color-undefined](https://user-images.githubusercontent.com/6458278/114345016-4638c700-9ba4-11eb-99bf-8f227064ef13.gif)

If I change this line to `color: colorObject?.color || value,` it works as I'd expect.

Is there a reason why the style prop should be `undefined` here in the presence of `colorObject?.slug`? 

Or are there CSS definitions for `.has-{color}-border-color` classes that I'm not seeing (or something else I'm missing)?

Thanks!

aaronrobertshaw replied:
There's a good chance I missed something here so I'll look into it. Thanks for testing.

> Is there a reason why the style prop should be undefined here?

This was `undefined` so that when using a named color it uses the CSS classes generated for the color from the current palette. If it stored the color value there the inline style overrules it. This could become a problem if the theme changes the color palette or the user switches themes.

aaronrobertshaw replied:
> Strangely, changing the text color modifies the border color, though I can't see why in the computed styles.

I have the suspicion that the theme sets the border color to the `currentColor` which would be determined by the text color. The inline border style works for me when a border style is set though.

ramonjd replied:
> This was undefined so that when using a named color it uses the CSS classes generated for the color from the current palette.

Thanks, that's what I suspected, was just scratching my head as to where those CSS class definitions were being (or should be) generated.

ramonjd replied:
Thanks for the clarification. Just dropping here to add that I had the TwentyTwentyOne theme activated. 

I could see the expected behaviour after switching to TT1 Blocks.

aaronrobertshaw replied:
I've [added the missing filter](https://github.com/WordPress/gutenberg/pull/30124/commits/7a2a25c44198bb39e912928ca9c27f1e03a8bba8) to force the inline styles for themes that don't load their color palettes in the editor. 

Testing TwentyTwentyOne, the appropriate border colors are now correctly shown in the editor. The CSS classes on the frontend still do not appear to be having any effect there though.

aaronrobertshaw replied:
After a rebase, the named color css classes do work for the TwentyTwentyOne theme. TwentyTwenty and some other older themes don't look to be qualifying for the generation of the css classes via the theme.json. I'm still looking into those.

aaronrobertshaw replied:
@nosolosw would you be able to point me in the right direction for getting the color styles generated from the theme.json color palette available for themes such as TwentyTwenty?

I believe the active theme currently has to satisfy [this check](https://github.com/WordPress/gutenberg/blob/158b115556a7502da7dc0bd88ffe28ad37569a74/lib/global-styles.php#L136) before the necessary styles are included.

oandregal replied:
`useEditorFeature( 'color.palette' )` will have the color palette, whether or not the theme has support for theme.json

aaronrobertshaw replied:
The issue I'm trying to solve here is some older themes don't qualify to have the named color CSS classes included. In other words they are missing `has-<color>-border-color`.

If the TT1 Blocks or TwentyTwenty Blocks themes are active, they get the generated color classes fine. Something older such as the TwentyTwenty (non block based) theme doesn't get these so while the block has the correct CSS class on the frontend the CSS class doesn't exist in the stylesheet.

oandregal replied:
Ah, I see. Themes that don't support theme.json won't have the preset classes automatically enqueued for them, they add them themselves. At some point in the past, we experimented with enqueuing the classes for these themes as well but heard reports that this wasn't expected, so we stopped doing it.

What I see is that named border-color classes are properly attached to the block (for any theme), and that's the only thing relevant for the hook. In terms of the global stylesheet, I also see them generated and enqueued in the embedded stylesheet `global-styles-inline-css` for themes with theme.json support, so everything is working as expected.

One thing we should do is to add this as a dev note saying: there's this new property so themes need to enqueue the classes for the border-colors as well. Only if this feature becomes stable for 5.8. I'll add the label so we can find them later.

Also ccing a few theme people FYI (re: `has-<color>-border-color` classes) ― @carolinan @aristath @scruffian @MaggieCabrera

aaronrobertshaw replied:
Thanks for the explanation. That helps my understanding 👍

aaronrobertshaw committed:
Add filter to force inline border color style in editor

This fixes issue where themes don't load their palettes in the editor.

Approach is the same as the colors block support that was missed on first pass updating the borders block support.

oandregal commented on lib/block-supports/border.php:
```diff
@@ -6,26 +6,36 @@
-	// Border width, style etc can be added in the future.
-	$has_border_radius_support = gutenberg_block_has_support( $block_type, array( '__experimentalBorder', 'radius' ), false );
+	// Determine if any border related features are supported.
+	$has_border_color_support = gutenberg_has_border_support( $block_type, 'color' );
```
Can we favor `gutenberg_block_has_support` over `gutenberg_has_border_support` (remove this one)?

aaronrobertshaw replied:
I can make that change, `gutenberg_block_has_support` didn't exist when I first created this and I didn't notice it after the rebase. Thanks for bringing it to my attention.

aaronrobertshaw replied:
✅ Done.

oandregal commented on lib/block-supports/border.php:
```diff
@@ -38,40 +48,102 @@ function gutenberg_register_border_support( $block_type ) {
 
-	// Border width, style etc can be added here.
+	// Border style.
+	if ( gutenberg_has_border_support( $block_type, 'style' ) ) {
```
Can we unnest these to be ?

```php
if (
  gutenberg_has_border_support( $block_type, 'radius' ) &&
  isset( $block_attributes['style']['border']['radius'] )
) { ... }
```

aaronrobertshaw replied:
✅ Done. Thanks.

oandregal commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -44,9 +50,37 @@ export function hasBorderSupport( blockName ) {
+ * @param {string|Object} blockType Block name or block type object.
+ * @return {boolean}                Whether the border feature is supported.
+ */
+export function hasBorderFeatureSupport( feature, blockType ) {
```
Can we merge `hasBorderSupport` and `hasBorderFeatureSupport` into a single function `hasBorderSupport( blockType, feature = 'all' )` or something along those lines?

oandregal replied:
Related https://github.com/WordPress/gutenberg/pull/30124/files?file-filters%5B%5D=.js#r612670345

aaronrobertshaw replied:
This has been refactored ✅ 

Thanks.

oandregal commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -24,15 +27,18 @@ export function BorderPanel( props ) {
 	return (
 		<InspectorControls>
-			<PanelBody title={ __( 'Border settings' ) }>
+			<PanelBody title={ __( 'Border settings' ) } initialOpen={ false }>
```
It seems like we can simplify quite a bit the code across this file and the `border-*.js` files if we do (pseudo-code):

```js
const hasSupport = hasBorderSupport( props );

const isColorSupported  = ! useEditorFeature( 'border.customColor' );
const isRadiusSupported = ! useEditorFeature( 'border.customRadius' );
const isStyleSupported  = ! useEditorFeature( 'border.customStyle' );
const isWidthSupported  = ! useEditorFeature( 'border.customWidth' );

if ( ! hasSupport || allAreDisabled ) {
	return null;
}
  

return (
	<PanelBody >
		isBorderSupported && <BorderStyleEdit />
		isWidthSupported && <BorderWidthEdit />
		isRadiusSupported && <BorderRadiusEdit />
		isColorSupported && <BorderColorEdit />
	</PanelBody>
);
```

aaronrobertshaw replied:
Thanks for the suggestion. I've made these changes ✅

oandregal commented:
I've reviewed the server hooks (blocks.php) and the client hook (block.json). Can continue with the review in the following days.

aaronrobertshaw committed:
Use gutenberg_block_has_support and unnest support checks

aaronrobertshaw committed:
Refactor to handle disabled/support checks in root border hook

oandregal commented on lib/block-supports/border.php:
```diff
@@ -6,26 +6,36 @@
-	$has_border_radius_support = gutenberg_block_has_support( $block_type, array( '__experimentalBorder', 'radius' ), false );
+	// Determine if any border related features are supported.
+	$has_border_color_support = gutenberg_block_has_support( $block_type, array( '__experimentalBorder', 'color' ) );
+	$has_border_support       =
```
It looks like we should check for `$has_border_support = gutenberg_block_has_support( $block_type, array( '__experimentalBorder' ) );` instead of the individual properties.

Otherwise, we may run into this situation:

- a block.json has `support.__experimentalBorder: true`.
- the client will [accept](https://github.com/WordPress/gutenberg/blob/da9b4ae772801c311112cf56fa35d1525490b3f5/packages/block-editor/src/hooks/style.js#L33-L33) that as support
- the server will not because it checks for the individual properties

aaronrobertshaw replied:
Thanks for catching this one. I've fixed that oversight.

oandregal commented:
Going to add the dev note label as per this conversation https://github.com/WordPress/gutenberg/pull/30124/files#r614248792

oandregal reviewed (approved):
I've left a note that I'd like addressed at https://github.com/WordPress/gutenberg/pull/30124/files#r614149424 Other than that it's working nicely and the code is ready.

Great work here. Thanks for your patience and persistence!

aaronrobertshaw committed:
Fix border support check

aaronrobertshaw commented:
> I've left a note that I'd like addressed at https://github.com/WordPress/gutenberg/pull/30124/files#r614149424 Other than that it's working nicely and the code is ready.

I've fixed up the support check code as requested. It should be pretty close now 🚢 

> Great work here. Thanks for your patience and persistence!

Team effort! I appreciate all the time you've spent reviewing this and other PRs 🙇

oandregal commented:
Added missed docs for these settings at https://github.com/WordPress/gutenberg/pull/31039 (updated: also for styles https://github.com/WordPress/gutenberg/pull/31040 )

aaronrobertshaw commented:
Added another PR for updating the `block-supports.md` reference guide - https://github.com/WordPress/gutenberg/pull/31069

oandregal commented:
We forgot to register the border color preset in the client as well, so the preset classes are not being generated in the site editor. Fix for this at https://github.com/WordPress/gutenberg/pull/31217

youknowriad commented:
I remove the dev note label because it's still experimental we should add the label to the PR that make this stable.
